From b4ca51f2185308bab2e496c6220634d88b57d121 Mon Sep 17 00:00:00 2001
From: Sagar Bussa <sagar.bussa@morsemicro.com>
Date: Mon, 10 Mar 2025 18:24:06 +1100
Subject: [PATCH 03/14] mac80211: tx s1g AP ecsa support

Update the AP-side MLME to consider short beacons in channel switch time
calculations, which includes an ECSA IE only in DTIM beacons. ECSA countdown
is updated only for DTIM beacons and not for short beacons.

This patch is S1G-specific and can have an impact on ECSA functionality of
other Wi-Fi radios, which delays channel switch operation based on the DTIM
interval. Impact on the ECSA functionality can be mitigated when combined with
'mac80211: mlme s1g ecsa support' and used on both APs and stations.
---
 net/mac80211/tx.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/net/mac80211/tx.c b/net/mac80211/tx.c
index 9142d748a6a7..fd3a92be02d7 100644
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -5039,11 +5039,11 @@ static void ieee80211_set_beacon_cntdwn(struct ieee80211_sub_if_data *sdata,
 
 static u8 __ieee80211_beacon_update_cntdwn(struct beacon_data *beacon)
 {
-	beacon->cntdwn_current_counter--;
-
 	/* the counter should never reach 0 */
 	WARN_ON_ONCE(!beacon->cntdwn_current_counter);
 
+	beacon->cntdwn_current_counter--;
+
 	return beacon->cntdwn_current_counter;
 }
 
@@ -5313,7 +5313,13 @@ ieee80211_beacon_get_ap(struct ieee80211_hw *hw,
 	u16 csa_off_base = 0;
 	int mbssid_len;
 
-	if (beacon->cntdwn_counter_offsets[0]) {
+	bool short_beacon = (vif->bss_conf.dtim_period > 1);
+
+	if (ap->ps.dtim_count > 0)
+		short_beacon = ((ap->ps.dtim_count-1) != 0);
+
+	 /* Do not count channel switch count for short beacons */
+	if (beacon->cntdwn_counter_offsets[0] && !short_beacon) {
 		if (!is_template)
 			ieee80211_beacon_update_cntdwn(vif, link->link_id);
 
-- 
2.34.1

